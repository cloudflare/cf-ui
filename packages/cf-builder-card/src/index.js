import cardActions from './cardActions';
import CardActionTypes from './CardActionTypes';
import CardBuilder from './CardBuilder';
import cardReducer from './cardReducer';

export { cardActions, CardActionTypes, CardBuilder, cardReducer };
