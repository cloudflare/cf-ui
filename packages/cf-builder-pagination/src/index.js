import PaginationBuilder from './PaginationBuilder';

export { PaginationBuilder };
