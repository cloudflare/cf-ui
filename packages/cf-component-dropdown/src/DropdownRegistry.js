export default class DropdownRegistry {
  constructor() {
    this._links = [];
    this._currentFocusIndex = -1;
  }

  addChild(child) {
    this._links.push(child);
  }

  removeChild(child) {
    this._links.splice(this._links.indexOf(child), 1);
  }

  _focusChildAt(index) {
    if (this._links[index]) {
      this._links[index].focus();
    }
  }

  focusNext() {
    this._focusChildAt(this._currentFocusIndex + 1);
  }

  focusPrev() {
    this._focusChildAt(this._currentFocusIndex - 1);
  }

  setFocusedChild(child) {
    this._currentFocusIndex = this._links.indexOf(child);
  }

  removeFocusedChild() {
    this._currentFocusIndex = -1;
  }
}
