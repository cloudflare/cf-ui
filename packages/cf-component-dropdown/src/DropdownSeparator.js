import React from 'react';
import { createComponent } from 'cf-style-container';

const styles = ({ theme }) => ({
  margin: '0.53333rem 0',
  borderTop: `1px solid ${theme.colorGrayLight}`
});

const DropdownSeparator = ({ className }) => (
  <div className={className} role="separator" />
);

export default createComponent(styles, DropdownSeparator);
