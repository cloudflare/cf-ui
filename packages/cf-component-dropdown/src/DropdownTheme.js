export default baseTheme => ({
  textAlign: 'initial'
});
