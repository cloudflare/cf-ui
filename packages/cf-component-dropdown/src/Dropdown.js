import React from 'react';
import PropTypes from 'prop-types';
import { canUseDOM } from 'exenv';
import { createComponent } from 'cf-style-container';

import DropdownRegistry from './DropdownRegistry';

const styles = ({ theme, align }) => ({
  position: 'absolute',
  zIndex: 1,
  minWidth: '10.66667rem',
  margin: '0.5em 0 0',
  padding: '0.33333rem 0',
  listStyle: 'none',
  background: theme.colorWhite,
  border: `1px solid ${theme.colorGrayLight}`,
  borderRadius: theme.borderRadius,
  boxShadow: '0 3px 10px rgba(0, 0, 0, 0.2)',

  left: align === 'left' ? 0 : 'initial',
  right: align === 'right' ? 0 : 'initial',

  animationName: {
    '0%': {
      display: 'none',
      opacity: 0
    },
    '1%': {
      display: 'block',
      opacity: 0,
      top: '80%'
    },
    '100%': {
      display: 'none',
      opacity: 1,
      top: '102%'
    }
  },
  animationDuration: '150ms',
  animationTimingFunction: 'ease-out',

  '&::before': {
    content: "''",
    display: 'block',
    position: 'absolute',
    bottom: '100%',
    border: 'solid transparent',
    borderWidth: '10px',
    borderTopWidth: 0,
    borderBottomColor: theme.colorWhite,
    left: align === 'left' ? '10px' : 'initial',
    right: align === 'right' ? '10px' : 'initial'
  }
});

class Dropdown extends React.Component {
  getChildContext() {
    return {
      dropdownRegistry: this.dropdownRegistry
    };
  }

  constructor(props, context) {
    super(props, context);
    this.dropdownRegistry = new DropdownRegistry();
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
    this.handleDocumentKeydown = this.handleDocumentKeydown.bind(this);
  }

  componentDidMount() {
    if (canUseDOM) {
      global.document.addEventListener('keydown', this.handleDocumentKeydown);
      global.document.addEventListener('click', this.handleDocumentClick);
    }
  }

  componentWillUnmount() {
    if (canUseDOM) {
      global.document.removeEventListener(
        'keydown',
        this.handleDocumentKeydown
      );
      global.document.removeEventListener('click', this.handleDocumentClick);
    }
  }

  handleDocumentKeydown(event) {
    const keyCode = event.keyCode;

    if (keyCode === 40) {
      // down
      event.preventDefault();
      this.dropdownRegistry.focusNext();
    } else if (keyCode === 38) {
      // up
      event.preventDefault();
      this.dropdownRegistry.focusPrev();
    } else if (keyCode === 27) {
      // esc
      this.props.onClose();
    }
  }

  handleDocumentClick() {
    this.props.onClose();
  }

  render() {
    return (
      <ul className={this.props.className} role="menu">
        {this.props.children}
      </ul>
    );
  }
}

Dropdown.propTypes = {
  onClose: PropTypes.func.isRequired,
  align: PropTypes.oneOf(['left', 'right']),
  children: PropTypes.node
};

Dropdown.defaultProps = {
  align: 'left'
};

Dropdown.childContextTypes = {
  dropdownRegistry: PropTypes.instanceOf(DropdownRegistry).isRequired
};

export default createComponent(styles, Dropdown);
