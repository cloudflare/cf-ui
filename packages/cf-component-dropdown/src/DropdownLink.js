import React from 'react';
import PropTypes from 'prop-types';
import Link from 'cf-component-link';
import { createComponent } from 'cf-style-container';

import DropdownRegistry from './DropdownRegistry';

const styled = ({ theme }) => ({
  position: 'relative',
  zIndex: 1,

  '&:first-child': {
    borderTopWidth: 0
  },

  '& > a': {
    display: 'block',
    padding: '0.2rem 1.06667rem',
    cursor: 'pointer',
    whiteSpace: 'nowrap',

    '&:hover': {
      color: theme.colorWhite,
      background: theme.color.marine
    },

    '&:focus': {
      color: theme.colorWhite,
      background: theme.color.sky,
      outline: '5px auto -webkit-focus-ring-color',
      outlineOffset: '-1px'
    },

    '&:active': {
      color: theme.colorWhite,
      background: theme.color.rain
    }
  }
});

class DropdownLink extends React.Component {
  constructor(props, context) {
    if (!props.to && !props.onClick) {
      throw new Error(
        '<DropdownLink/> requires either a `to` or `onClick` prop'
      );
    }

    super(props, context);
    this.dropdownRegistry = context.dropdownRegistry;

    this.handleLinkBlur = this.handleLinkBlur.bind(this);
    this.handleLinkFocus = this.handleLinkFocus.bind(this);
    this.focus = this.focus.bind(this);
  }

  componentDidMount() {
    this.dropdownRegistry.addChild(this);
  }

  componentWillUnmount() {
    this.dropdownRegistry.removeChild(this);
  }

  focus() {
    this.link.focus();
  }

  handleLinkFocus() {
    this.dropdownRegistry.setFocusedChild(this);
  }

  handleLinkBlur() {
    this.dropdownRegistry.removeFocusedChild();
  }

  render() {
    return (
      <li className={this.props.className} role="menuitem">
        <Link
          ref={node => this.link = node}
          to={this.props.to}
          onClick={this.props.onClick}
          onFocus={this.handleLinkFocus}
          onBlur={this.handleLinkBlur}
        >
          {this.props.children}
        </Link>
      </li>
    );
  }
}

DropdownLink.propTypes = {
  to: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node
};

DropdownLink.contextTypes = {
  dropdownRegistry: PropTypes.instanceOf(DropdownRegistry).isRequired
};

export default createComponent(styled, DropdownLink);
