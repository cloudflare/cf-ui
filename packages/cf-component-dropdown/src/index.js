import { applyTheme } from 'cf-style-container';

import DropdownUnstyled from './Dropdown';
import DropdownTheme from './DropdownTheme';
import DropdownLinkUnstyled from './DropdownLink';
import DropdownLinkTheme from './DropdownLinkTheme';
import DropdownSeparatorUnstyled from './DropdownSeparator';
import DropdownSeparatorTheme from './DropdownTheme';

const Dropdown = applyTheme(DropdownUnstyled, DropdownTheme);
const DropdownLink = applyTheme(DropdownLinkUnstyled, DropdownLinkTheme);
const DropdownSeparator = applyTheme(
  DropdownSeparatorUnstyled,
  DropdownSeparatorTheme
);

export {
  Dropdown,
  DropdownTheme,
  DropdownUnstyled,
  DropdownLink,
  DropdownLinkTheme,
  DropdownLinkUnstyled,
  DropdownSeparator,
  DropdownSeparatorTheme,
  DropdownSeparatorUnstyled
};
