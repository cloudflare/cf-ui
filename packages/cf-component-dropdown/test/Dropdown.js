import React from 'react';
import renderer from 'react-test-renderer';
import { render, unmountComponentAtNode } from 'react-dom';
import { felaSnapshot, felaTestContext } from 'cf-style-provider';
import jsdom from 'jsdom';
import { mount } from 'enzyme';

import { Dropdown, DropdownLink } from '../../cf-component-dropdown/src/index';

test('should render', () => {
  const snapshot = felaSnapshot(
    <Dropdown onClose={() => {}}>
      Dropdown
    </Dropdown>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render with align', () => {
  const snapshot = felaSnapshot(
    <Dropdown align="right" onClose={() => {}}>
      Dropdown
    </Dropdown>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

describe('interactions', () => {
  beforeEach(function() {
    const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
    global.document = doc;
    global.window = doc.defaultView;
    global.root = global.document.createElement('div');
    global.document.body.appendChild(global.root);
  });

  afterEach(function() {
    unmountComponentAtNode(global.root);
    global.document.body.removeChild(global.root);
    delete global.root;
  });

  test('should call `onClose` when the `esc` key is pressed', () => {
    var called = false;
    var onClose = function() {
      called = true;
    };

    var dom = mount(felaTestContext(<Dropdown onClose={onClose} />));

    // Fela renders a Stateless component so we can't just use `.getNode()` but
    // we know that the Dropdown we want is the second element in the list.
    const instance = dom.find('Dropdown').getNodes()[1];

    instance.handleDocumentKeydown({ keyCode: 27 });
    expect(called).toBeTruthy();
  });

  test('should be able to navigate with the up and down keys', () => {
    const down = { keyCode: 40, preventDefault() {} };
    const up = { keyCode: 38, preventDefault() {} };

    const getFocusedHref = () =>
      global.document.activeElement.getAttribute('href');

    const dom = mount(
      felaTestContext(
        <Dropdown onClose={() => {}}>
          <DropdownLink to="one" />
          <DropdownLink to="two" />
          <DropdownLink to="three" />
        </Dropdown>
      )
    );
    const instance = dom.find('Dropdown').getNodes()[1];

    expect(getFocusedHref()).toBe(null);

    instance.handleDocumentKeydown(down);
    setTimeout(() => expect(getFocusedHref()).toBe('one'), 0);

    instance.handleDocumentKeydown(down);
    setTimeout(() => expect(getFocusedHref()).toBe('two'), 0);

    instance.handleDocumentKeydown(down);
    instance.handleDocumentKeydown(down);
    setTimeout(() => expect(getFocusedHref()).toBe('three'), 0);

    instance.handleDocumentKeydown(up);
    setTimeout(() => expect(getFocusedHref()).toBe('two'), 0);

    const two = global.document.activeElement;

    instance.handleDocumentKeydown(up);
    instance.handleDocumentKeydown(up);
    setTimeout(() => expect(getFocusedHref()).toBe('one'), 0);

    two.focus();
    instance.handleDocumentKeydown(down);
    setTimeout(() => expect(getFocusedHref()).toBe('three'), 0);
  });
});
