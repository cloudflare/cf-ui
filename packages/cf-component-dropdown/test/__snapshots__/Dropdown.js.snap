// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<ul
  className="a b c d e f g h i j k l m n o p q r s t u v w x y"
  role="menu"
>
  Dropdown
</ul>
`;

exports[`should render 2`] = `
"
@-webkit-keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

@-moz-keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

@keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

.a {
  position: absolute
}

.b {
  z-index: 1
}

.c {
  min-width: 10.66667rem
}

.d {
  margin: 0.5em 0 0
}

.e {
  padding: 0.33333rem 0
}

.f {
  list-style: none
}

.g {
  background: #fff
}

.h {
  border: 1px solid #dedede
}

.i {
  border-radius: 2px
}

.j {
  box-shadow: 0 3px 10px rgba(0, 0, 0, 0.2)
}

.k {
  left: initial
}

.l {
  right: initial
}

.m {
  animation-name: k1
}

.n {
  animation-duration: 150ms;
  -webkit-animation-duration: 150ms
}

.o {
  animation-timing-function: ease-out;
  -webkit-animation-timing-function: ease-out
}

.p::before {
  content: ''
}

.q::before {
  display: block
}

.r::before {
  position: absolute
}

.s::before {
  bottom: 100%
}

.t::before {
  border: solid transparent
}

.u::before {
  border-width: 10px
}

.v::before {
  border-top-width: 0px
}

.w::before {
  border-bottom-color: #fff
}

.x::before {
  left: initial
}

.y::before {
  right: initial
}
"
`;

exports[`should render with align 1`] = `
<ul
  className="a b c d e f g h i j k l m n o p q r s t u v w x y"
  role="menu"
>
  Dropdown
</ul>
`;

exports[`should render with align 2`] = `
"
@-webkit-keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

@-moz-keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

@keyframes k1 {
  0% {
    display: none;
    opacity: 0
  }

  1% {
    display: block;
    opacity: 0;
    top: 80%
  }

  100% {
    display: none;
    opacity: 1;
    top: 102%
  }
}

.a {
  position: absolute
}

.b {
  z-index: 1
}

.c {
  min-width: 10.66667rem
}

.d {
  margin: 0.5em 0 0
}

.e {
  padding: 0.33333rem 0
}

.f {
  list-style: none
}

.g {
  background: #fff
}

.h {
  border: 1px solid #dedede
}

.i {
  border-radius: 2px
}

.j {
  box-shadow: 0 3px 10px rgba(0, 0, 0, 0.2)
}

.k {
  left: initial
}

.l {
  right: 0px
}

.m {
  animation-name: k1
}

.n {
  animation-duration: 150ms;
  -webkit-animation-duration: 150ms
}

.o {
  animation-timing-function: ease-out;
  -webkit-animation-timing-function: ease-out
}

.p::before {
  content: ''
}

.q::before {
  display: block
}

.r::before {
  position: absolute
}

.s::before {
  bottom: 100%
}

.t::before {
  border: solid transparent
}

.u::before {
  border-width: 10px
}

.v::before {
  border-top-width: 0px
}

.w::before {
  border-bottom-color: #fff
}

.x::before {
  left: initial
}

.y::before {
  right: 10px
}
"
`;
